### ISSUE TITLE
Does the stateful_lstm.py example make sense?

### ISSUE THREAD

**Issue by yaringal** [fchollet/keras#1820]
I was going over the example in:
https://github.com/fchollet/keras/blob/master/examples/stateful_lstm.py @fchollet
and I'm a bit confused. The RNN assumes size (nb x 1 x 1). When cos (of size N x 1 x 1) is given as input, model._fit will randomly sample different mini-batches from the input at each iteration. How is the temporal dependency kept? I'm also confused about the propagation of derivatives with stateful RNNs -  how are the derivatives wrt the internal states propagated back in time?

**Comment by fchollet** [comment]
> I'm also confused about the propagation of derivatives with stateful RNNs - how are the derivatives wrt the internal states propagated back in time?

They're not. "Stateful" only means that the final states from one batch are reused as the initial states for the next batch.

I don't understand your first question.

**Comment by yaringal** [comment]
Cool. In the first question I meant that the final states from the previous batch will have nothing to do with the current batch if they are sampled at random. Or am I missing something?

**Comment by NasenSpray** [comment]
@fchollet It actually doesn't:
1. `shuffle=False` is missing in call to `fit()` 
2. training data is not shuffled in regard to `batch_size` (`k`th timestep for sequence `s` should be at `[s+k*batch_size, ...]`)

(2) can be fixed by inserting after [line 41](https://github.com/fchollet/keras/blob/master/examples/stateful_lstm.py#L41):
```
assert cos.shape[0] % batch_size == 0
cos = cos.reshape(batch_size, -1, cos.shape[1], cos.shape[2])
cos = cos.transpose(1, 0, 2, 3)
cos = cos.reshape(-1, cos.shape[2], cos.shape[3])
```

Before:
> Epoch 0 / 25
Epoch 1/1
50000/50000 [==============================] - 22s - loss: 339.1427
Epoch 1 / 25
Epoch 1/1
50000/50000 [==============================] - 22s - loss: 149.8113

Fixed:
> Epoch 0 / 25
Epoch 1/1
50000/50000 [==============================] - 22s - loss: 79.9879
Epoch 1 / 25
Epoch 1/1
50000/50000 [==============================] - 23s - loss: 9.7246

**Comment by lukovkin** [comment]
@NasenSpray I agree with 1st point (shuffle=False), but would you be so kind to elaborate on your additions on shuffle in regard to batch_size?
I've added your insertion to the code and it effectively shuffles the input sequence the way when 1st element of the initial 'cos' will be followed by the 2000th element, 2nd element - by the 2001st element, and so on. It is strange and results in the very strange output:
![image](https://cloud.githubusercontent.com/assets/5587574/13361674/304bbf16-dcd1-11e5-9fef-84c0543c4182.png)

May be you could explain a little bit more?

**Comment by fchollet** [comment]
Yes, shuffle should be set to False. Thanks for spotting this.

**Comment by lukovkin** [comment]
@fchollet credits go to @NasenSpray actually ;)

**Comment by NasenSpray** [comment]
@lukovkin think about how data is fetched and what that means for stateful RNNs. Example with batch size 2:

![image](http://i.imgur.com/3KkX0V1.png)

As you can see, the sequences must be interleaved in such a way that the individual "lanes" in a mini-batch get consecutive timesteps of their sequence. The code cuts the data in `len/batch_size` chunks and interleaves them, but I forgot about the plots and that's why it looks so messed up.

**Comment by lukovkin** [comment]
@NasenSpray Sorry, I don't get your notation wrt LSTM(n) - what does LSTM(0) and LSTM(1) exactly mean and why LSTM is spread across the batches?
As I could see from the docs and code,

> You can set RNN layers to be 'stateful', which means that the states
        computed for the samples in one batch will be reused as initial states
        for the samples in the next batch.

And in the code it also just sets initial_states equal to states.
So as far as I understand, LSTM itself is not being spanned across the batches.

**Comment by 9aradox** [comment]
I really think something about stateful is not working. My biggest confirmation of this is the given example (stateful_lstm.py) works exactly the same if you set stateful=False

It would be great if there was some example toy problem that needed state and showed how it works. I am trying to understand the code but theano debugging still makes my brain explode.

I was trying to modify the text generation to  emulate the behavior of this code base (karpathy/min-char-rnn.py). But even though I think I have everything working it only really works when timesteps are large enough. At 1 timestep it does not converge beyond basic 1 step prediction.

**Comment by bhy** [comment]
> I really think something about stateful is not working. My biggest confirmation of this is the given example (stateful_lstm.py) works exactly the same if you set stateful=False

@9aradox this is because `fit` and `predict` are called with the entire sequence, so stateful or not the result would be the same.

**Comment by bhy** [comment]
From the Keras FAQ (http://keras.io/faq/#how-can-i-use-stateful-rnns) it says:

1. all batches have the same number of samples
2. If `X1` and `X2` are successive batches of samples, then `X2[i]` is the follow-up sequence to `X1[i]`, for every `i`.

While `model.fit` will make mini-batches that does not follow these assumptions (especially 2.). To illustrate the problem, if we change the input sequence length to be non-dividable by `batch_size`:

    cos = gen_cosine_amp(xn=507)

training will crash.

**Comment by 9aradox** [comment]
I am not sure what you are saying. I think I understand the batching issues. I don't have any problem running code it is just if you have a problem that needs state and you trying to use 1 timestep then it just doesn't converge. 

I do find the number 2 restriction confusing but I am pretty sure with 1 timestep like this example it doesn't matter.

None of this explains why the stateful example in keras works fine when you set stateful=False. This seems to invalidate it as an example. Also it does not matter if you do the prediction one item at a time. This code also works fine with stateful = False in both layers.

#WARNING This loop takes forever.
print('Predicting')
predicted_output = []
x = np.zeros((batch_size, 1, 1))
for c in cos:
    x [0] = c
    prediction = model.predict(x, batch_size=1)
    #print (prediction)
    predicted_output.append(prediction[0][0])

**Comment by lukovkin** [comment]
@9aradox I had stateful working with batch_size=1, but not for cos in example, but for financial TS data.
It is interesting that val loss with stateful and shuffle=True was better than for stateful=True and shuffle=False, which doesn't make a lot of sense.

**Comment by gvessere** [comment]
I think a reason why the example still works in the stateful=false case, might have to do with the fact that the period on the cosine function is 25, the same as the batch size, this would equate to predicting a bunch of exponentials going to zero. I haven't tried but worth mentioning. I think @NasenSpray had it right by reshuffling the series such that consecutive batches process the series in order as in the doc.

**Comment by sadeghmir** [comment]
> I think a reason why the example still works in the stateful=false case, might have to do with the fact that the period on the cosine function is 25, the same as the batch size, this would equate to predicting a bunch of exponentials going to zero.

You would think the period is 25 because of the argument `period` in `gen_cosine_amp`, but it is in fact close to 1000 due to the way that argument is used:

`cos[i, 0, 0] = amp * np.cos(idx / (2 * np.pi * period))` 

I think the actual reason for why `stateful=False` still gets "good" results is that the model tries to predict only one step ahead, which is a very easy task because of the data. To see how easy it is, consider the simple baseline where your prediction of one step ahead is exactly the same as the current value. This simple baseline, surprisingly, achieves MSE=0.0203 which is way better than anything the current model achieves. To really test the effect of statefulness, one should set the look-ahead (prediction point ahead of current time) to something like 100 to make the problem more challenging. When you do that, be careful to to use 

`expected_output[i, 0] = cos[i + lahead, 0, 0]`

At line 44, because for some reason the current code tries to predict the average of look-ahead values. I did all this and I got an MSE of 35 and 243 for stateful and stateless models respectively, which clearly shows the benefit of statefulness. I think the example should be modified as well.

**Comment by sadeghmir** [comment]
@NasenSpray 
Your solution works when (as you said yourself) the labels are also rearranged. Here is the code for those interested. Add all of these after line 48:
```
print('Rearranging the data to satisfy stateful requirements')
cos = cos.reshape(batch_size, -1, cos.shape[1], cos.shape[2])
cos = cos.transpose(1, 0, 2, 3)
cos = cos.reshape(-1, cos.shape[2], cos.shape[3])    
expected_output = expected_output.reshape(batch_size, -1, expected_output.shape[1])
expected_output = expected_output.transpose(1, 0, 2)
expected_output = expected_output.reshape(-1, expected_output.shape[2])
``` 

The fixed version will get an MSE close to 2 for lookahead=100 in just 3 epochs which is pretty good (compare to the results I reported in my previous comment).

**Comment by yongduek** [comment]
@sadeghmir 
I recently have been thinking of this problem and you provided exactly right answer. Thanks a log.

A small observation is below.

With the modification
  ``     expected_output[i, 0] = cos[i + lahead, 0, 0]``
I had:
``stateful=False``  resulted in loss 241
``stateful=True``  resulted in loss 53

Then, in the hope of making the learning faster, I changed the batch_size to 2500 from 25.
Thanks to nvidia titan X, it became much faster but the result was quite different, the loss was 322. However, the output waveform was much better then the former two cases. So, I change the `epochs=250` from `25`, which resulted in almost all the same wave form with loss 166!
My conclusion is that the magnitude of the loss is somewhat unrelated to the overall waveform reconstruction.

**Comment by FabianFalck** [comment]
How does the statement 

> If x1 and x2 are successive batches of samples, then x2[i] is the follow-up sequence to x1[i], for every i.

make sense in the use case of a time series? Instead of having a dependency between x1[i] and x2[i], one has a dependency between x1[i] and x1[i+1], so between two adjacent samples within one batch. I strongly believe that the way the data has to be modelled for LSTMs in Keras is misleading, even in other contexts. I cannot see how this way would make sense.
